test/unit/gateways/realex_test.rb

```diff
@@ -51,6 +51,41 @@ def setup
     @amount = 100
   end
 
+  def test_initialize_sets_refund_and_credit_hashes
+    refund_secret = 'refund'
+    rebate_secret = 'rebate'
+
+    options = {
+      login: @login,
+      password: @password,
+      rebate_secret: rebate_secret,
+      refund_secret: refund_secret
+    }
+
+    gateway = RealexGateway.new(options)
+
+    set_options = gateway.instance_variable_get(:@options)
+
+    assert set_options[:refund_hash] == Digest::SHA1.hexdigest(rebate_secret)
+    assert set_options[:credit_hash] == Digest::SHA1.hexdigest(refund_secret)
+  end
+
+  def test_initialize_with_nil_refund_and_rebate_secrets
+    options = {
+      login: @login,
+      password: @password,
+      rebate_secret: nil,
+      refund_secret: nil
+    }
+
+    gateway = RealexGateway.new(options)
```

Comment:
Should this made into an `assert` statement since we just want to test whether we can successfully create a new gateway object when passing in `nil` for the secrets?

Reply:
Another way to think about unit tests is that they define the expected behavior of the method. Failing to create a gateway object (due to a type error) when passing in ```nil``` for the secrets led us to add tests here, but the test more generally tests all of the expected behavior. So from a definitional standpoint, someone reading this test can see that a key for the each option won't be added to the options hash (rather than, say, a key being added and set to ```nil```). The asserts themselves also enforce that behavior, so there would be a failure if someone refactored ```initialize``` later such that it that added keys with ```nil``` or any other value.

Reply:
This seems to be an answer to my other question.

This question was simply whether it be good to add an `assert` at the beginning of this line to create a test that confirms that the gateway is successfully created since that was the specific behavior that was breaking previously.

Reply:
Hmm. Nothing comes to mind for me as far as a case where ```initialize``` could return ```nil```. In the scenario where we had it breaking before, you would just have the test fail due to the type error, so the ```assert``` would never get tested.

Reply:
Adding the `assert` in at this point does return the `TypeError` and fails the test:
![image](https://user-images.githubusercontent.com/42787645/62485790-b1a18300-b78b-11e9-86be-294d0f16563c.png)

![image](https://user-images.githubusercontent.com/42787645/62485819-cc73f780-b78b-11e9-8fd2-18cfed6f72c1.png)

Reply:
For this test, for the original bug, you get that exception thrown whether the assert is added or not. The assert doesn't capture the exception, so it doesn't fail the assert it's just a normal unhandled exception getting surfaced.

Unless `RealexGateway.new` is redefined to do something very non-standard, it won't return `nil`. The constructor will always return some kind of object, or raise an error, in the common case.

I don't think the extra assert would add value in this specific case.

Reply:
Gotcha, thanks for the clarifications @bayprogrammer!